pmlc/dialect/tile/transforms/padding.cc

```diff
@@ -0,0 +1,213 @@
+// Copyright 2019, Intel Corporation
+
+#include "mlir/IR/AffineExprVisitor.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/Pass/Pass.h"
+
+#include "pmlc/dialect/eltwise/ir/ops.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/tile/transforms/passes.h"
+
+namespace pmlc::dialect::tile {
+
+using mlir::AffineMap;
+
+struct BoundRange {
+  int64_t min;
+  int64_t max;
+  explicit BoundRange(int64_t val) : min(val), max(val) {}
+  BoundRange(int64_t min, int64_t max) : min(min), max(max) {}
+  BoundRange operator+(const BoundRange &rhs) const {
+    return {min + rhs.min, max + rhs.max};
+  }
+  BoundRange operator*(const BoundRange &rhs) const {
+    return {std::min(std::min(min * rhs.min, max * rhs.min),
+                     std::min(min * rhs.max, max * rhs.max)),
+            std::max(std::max(min * rhs.min, max * rhs.min),
+                     std::max(min * rhs.max, max * rhs.max))};
+  }
+};
+
+class ComputeRangeVisitor
+    : public mlir::AffineExprVisitor<ComputeRangeVisitor, BoundRange> {
+public:
+  explicit ComputeRangeVisitor(std::vector<BoundRange> dimRanges)
+      : dimRanges(std::move(dimRanges)) {}
+
+  BoundRange visitMulExpr(mlir::AffineBinaryOpExpr expr) {
+    return visit(expr.getLHS()) * visit(expr.getRHS());
+  }
+  BoundRange visitAddExpr(mlir::AffineBinaryOpExpr expr) {
+    return visit(expr.getLHS()) + visit(expr.getRHS());
+  }
+  BoundRange visitDimExpr(mlir::AffineDimExpr expr) {
+    return dimRanges[expr.getPosition()];
+  }
+  BoundRange visitSymbolExpr(mlir::AffineSymbolExpr expr) {
+    llvm_unreachable("Invalid symbol in ComputeRangeVisitor");
+  }
+  BoundRange visitConstantExpr(mlir::AffineConstantExpr expr) {
+    return BoundRange(expr.getValue());
+  }
+  BoundRange visitModExpr(mlir::AffineBinaryOpExpr expr) {
+    llvm_unreachable("Invalid mod in ComputeRangeVisitor");
+  }
+  BoundRange visitCeilDivExpr(mlir::AffineBinaryOpExpr expr) {
+    llvm_unreachable("Invalid ceil in ComputeRangeVisitor");
+  }
+  BoundRange visitFloorDivExpr(mlir::AffineBinaryOpExpr expr) {
+    llvm_unreachable("Invalid floor in ComputeRangeVisitor");
+  }
+
+private:
+  std::vector<BoundRange> dimRanges;
+};
+
+llvm::SmallVector<BoundRange, 4>
+computeBounds(AffineMap access, AffineMap lower, AffineMap upper) {
```

Comment:
This is slightly confusing because we already have a pass called "compute bounds" and presumably this function doesn't do the same work.